README.md
# express-response-formatter

[![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/aofleejay/express-response-formatter/blob/master/LICENSE.md) [![npm](https://img.shields.io/npm/v/express-response-formatter.svg)](https://www.npmjs.com/package/express-response-formatter) [![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg)](https://github.com/prettier/prettier) [![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg)](http://makeapullrequest.com) [![Build Status](https://travis-ci.org/aofleejay/express-response-formatter.svg?branch=master)](https://travis-ci.org/aofleejay/express-response-formatter) [![Coverage Status](https://coveralls.io/repos/github/aofleejay/express-response-formatter/badge.svg?branch=master)](https://coveralls.io/github/aofleejay/express-response-formatter?branch=master)

Better way to format Express response

## How It Works

- You can use response with readable name like `res.formatter.ok` for 200 ok or `res.formatter.badRequest` for 400 bad request.
- It will format your response in two ways `success` and `error`.
  - If response is `2xx, 3xx` return response under object key `data`.
  - If response is `4xx, 5xx` return response under object key `error`.
- You can pass metadata as second parameter and it's will present under `meta` object key.

![Usage](usage.gif)

## Installation

```
npm install express-response-formatter --save
```

## Quick Start

Response for 200 Ok.

```js
import app from 'express'
import { responseEnhancer } from 'express-response-formatter'

const app = express()

// Add formatter functions to "res" object via "responseEnhancer()"
app.use(responseEnhancer())

app.get('/success', (req, res) => {
  const users = [{ name: 'Dana Kennedy' }, { name: 'Warren Young' }]

  // It's enhance "res" with "formatter" which contain formatter functions
  res.formatter.ok(users)
})

app.listen(3000, () => console.log('Start at http://localhost:3000'))
```

Result

```json
HTTP/1.1 200 Ok
{
  "data": [
    {
      "name": "Dana Kennedy"
    },
    {
      "name": "Warren Young"
    }
  ]
}
```

## More usages

Response for 200 Ok with `meta field`

```js
app.get('/success-with-meta', (req, res) => {
  const users = [{ name: 'Dana Kennedy' }, { name: 'Warren Young' }]

  const meta = {
    total: 2,
    limit: 10,
    offset: 0,
  }

  res.formatter.ok(users, meta)
})
```

```json
HTTP/1.1 200 Ok
{
  "meta": {
    "total": 2,
    "limit": 10,
    "offset": 0,
  },
  "data": [
    {
      "name": "Dana Kennedy"
    },
    {
      "name": "Warren Young"
    }
  ]
}
```

Response for 400 Bad Request with "error"

```js
app.get('/bad-request', (req, res) => {
  const errors = [
    { detail: 'Field id is required.' },
    { detail: 'Field foo is required.' },
  ]

  res.formatter.badRequest(errors)
})
```

```json
HTTP/1.1 400 Bad Request
{
  "error": [
    {
      "detail": "Field id is required."
    },
    {
      "detail": "Field foo is required."
    }
  ]
}
```

## APIs

| METHOD                                         | STATUS CODE |
| ---------------------------------------------- | ----------- |
| res.formatter.ok(data, meta?)                  | 200         |
| res.formatter.created(data, meta?)             | 201         |
| res.formatter.accepted(data, meta?)            | 202         |
| res.formatter.noContent(data, meta?)           | 204         |
| res.formatter.badRequest(errors, meta)         | 400         |
| res.formatter.unauthorized(errors, meta)       | 401         |
| res.formatter.forbidden(errors, meta)          | 403         |
| res.formatter.notFound(errors, meta)           | 404         |
| res.formatter.methodNotAllowed(errors, meta)   | 405         |
| res.formatter.timeout(errors, meta)            | 408         |
| res.formatter.conflict(errors, meta)           | 409         |
| res.formatter.unprocess(errors, meta)          | 422         |
| res.formatter.tooManyRequests(errors, meta)    | 429         |
| res.formatter.serverError(errors, meta)        | 500         |
| res.formatter.badGateway(errors, meta)         | 502         |
| res.formatter.serviceUnavailable(errors, meta) | 503         |
| res.formatter.gatewayTimeout(errors, meta)     | 504         |


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [2.0.2] - 2020-08-12

### Added

- Add auto complete for TypeScript.

## [2.0.1] - 2020-08-11

### Added

- Add accepted, timeout, conflict and tooManyRequests method. ([#8](https://github.com/aofleejay/express-response-formatter/pull/8))

## [2.0.0] - 2020-04-28

### Changed

- Use named export to prevent interop.
- Use 'error' key instead of 'errors' for more flexibility.
- Can pass meta as second patameter for error response.

## [1.2.7] - 2020-03-06

### Changed

- Remove ignore lib but manualy publish.

## [1.2.6] - 2020-03-06

### Changed

- Remove ignore lib.

## [1.2.5] - 2020-03-06

### Changed

- Fix CI build.

## [1.2.4] - 2020-03-06

### Changed

- Move from babel to tsc.

## [1.2.3] - 2020-03-05

### Changed

- Move build script to before_script in travis.yml.
- Update example folder.

## [1.2.2] - 2020-03-05

### Changed

- Remove lib folder in gitignore.

## [1.2.1] - 2020-03-05

### Changed

- Move build script to before_script in travis.yml.

## [1.2.0] - 2020-03-05

### Added

- Support TypeScript.
- Add CI.

## [1.1.0] - 2019-06-04

### Added

- Use prettier.
- Add 2xx, 4xx test.
- Add accepted, timeout, conflict and tooManyRequests method.

### Changed

- Upgrade to babel 7.

## [1.0.0] - 2017-06-20

### Added

- Format JSON response by enhance Express response object.


src/methods.ts
type Methods =
  | 'ok'
  | 'created'
  | 'accepted'
  | 'noContent'
  | 'badRequest'
  | 'unauthorized'
  | 'forbidden'
  | 'notFound'
  | 'methodNotAllowed'
  | 'timeout'
  | 'conflict'
  | 'unprocess'
  | 'tooManyRequests'
  | 'serverError'
  | 'badGateway'
  | 'serviceUnavailable'
  | 'gatewayTimeout'

interface Method {
  name: Methods
  code: string
  message: string
  isSuccess: boolean
}

const methods: Method[] = [
  {
    name: 'ok',
    code: '200',
    message: 'OK',
    isSuccess: true,
  },
  {
    name: 'created',
    code: '201',
    message: 'Created',
    isSuccess: true,
  },
  {
    name: 'accepted',
    code: '202',
    message: 'Accepted',
    isSuccess: true,
  },
  {
    name: 'noContent',
    code: '204',
    message: 'No Content',
    isSuccess: true,
  },
  {
    name: 'badRequest',
    code: '400',
    message: 'Bad Request',
    isSuccess: false,
  },
  {
    name: 'unauthorized',
    code: '401',
    message: 'Unauthorized',
    isSuccess: false,
  },
  {
    name: 'forbidden',
    code: '403',
    message: 'Forbidden',
    isSuccess: false,
  },
  {
    name: 'notFound',
    code: '404',
    message: 'Not Found',
    isSuccess: false,
  },
  {
    name: 'methodNotAllowed',
    code: '405',
    message: 'Method Not Allowed',
    isSuccess: false,
  },
  {
    name: 'timeout',
    code: '408',
    message: 'Timeout',
    isSuccess: false,
  },
  {
    name: 'conflict',
    code: '409',
    message: 'Conflict',
    isSuccess: false,
  },
  {
    name: 'unprocess',
    code: '422',
    message: 'Unprocessable Entity',
    isSuccess: false,
  },
  {
    name: 'tooManyRequests',
    code: '429',
    message: 'Too Many Requests',
    isSuccess: false,
  },
  {
    name: 'serverError',
    code: '500',
    message: 'Internal Server Error',
    isSuccess: false,
  },
  {
    name: 'badGateway',
    code: '502',
    message: 'Bad Gateway',
    isSuccess: false,
  },
  {
    name: 'serviceUnavailable',
    code: '503',
    message: 'Service Unavailable',
    isSuccess: false,
  },
  {
    name: 'gatewayTimeout',
    code: '504',
    message: 'Gateway Timeout',
    isSuccess: false,
  },
]

export { Method, Methods }
export default methods
